_LowLevelCode/cpp/cpp_communicator/serialiser/c_deserialise.cpp

```diff
@@ -0,0 +1,363 @@
+          std::vector<mxArray*> results(2);
+          std::vector<mxArray*> input {mxName, mxData};
+          mexCallMATLAB(2, results.data(), 2, input.data(), "c_hlp_deserialise_object_self");
+          output = results[0];
+          memPtr += mxGetScalar(results[1]);
+          mxDestroyArray(mxName);
+          mxDestroyArray(mxData);
+        }
+        break;
+      case SAVEOBJ:
+        {
+          mxArray* mxName = mxCreateString(name.data());
+          mxArray* conts = deserialise(data, size, 1);
+          std::vector<mxArray*> input {mxName, conts};
+          mexCallMATLAB(1, &output, 2, input.data(), "c_hlp_deserialise_object_loadobj");
+          mxDestroyArray(conts);
+          mxDestroyArray(mxName);
+        }
+        break;
+      case STRUCTED:
+          {
+            output = deserialise(data, size, 1);
+            mxSetClassName(output, name.data());
+          }
+        break;
+      }
+
+    }
+    break;
+
+  case STRUCT:
+    {
+      uint32_t nFields;
+      deser(data, &nFields, types_size[UINT32]);
+
+      std::vector<uint32_t> fNameLens(nFields);
+      deser(data, fNameLens, nFields*types_size[UINT32]);
+
+      std::vector<std::vector<char>> fNames(nFields);
+      for (int field=0; field < nFields; field++) {
+        fNames[field] = std::vector<char>(fNameLens[field]+1);
+        fNames[field][fNameLens[field]] = 0;
+        deser(data, fNames[field], fNameLens[field]*types_size[CHAR]);
+      }
+
+      output = mxCreateStructArray(nDims, dims, nFields, (const char**) fNames.data());
+      if (nFields == 0) break;
+
+      mxArray* cellData = deserialise(data, size, 1);
+
+      for (int obj=0, elem=0; obj < nElem; obj++) {
+        for (int field=0; field < nFields; field++, elem++) {
+          mxArray* cellElem = mxGetCell(cellData, elem);
+          mxSetFieldByNumber(output, obj, field, cellElem);
+        }
+      }
+
+    }
+    break;
+
+  case CELL:
+    {
+      output = mxCreateCellArray(nDims, dims);
+      for (mwIndex i = 0; i < nElem; i++) {
+        mxSetCell(output, i, deserialise(data, size, 1));
+      }
+
+    }
+    break;
+  }
+
+  // Avoid making plhs persistent
+  if (recursed) {
+    mexMakeArrayPersistent(output);
+  }
+  return output;
+}
+
+
+
+/* The gateway routine. */
+void mexFunction(int nlhs, mxArray* plhs[], int nrhs, const mxArray* prhs[] ) {
+
+#if defined(_LP64) || defined (_WIN64)
+#ifdef MX_COMPAT_32
+  for (i=0; i<nrhs; i++)  {
+    if (mxIsSparse(prhs[i])) {
+      mexErrMsgIdAndTxt("MATLAB:c_serialise:NoSparseCompat",
+                        "MEX-files compiled on a 64-bit platform that use sparse array functions "
+                        "need to be compiled using -largeArrayDims.");
+    }
+  }
+#endif
+#endif
+
+  if (nlhs > 1) mexErrMsgIdAndTxt("MATLAB:c_serialise:badLHS", "Bad number of LHS arguments in c_serialise");
+  if (nrhs != 1) mexErrMsgIdAndTxt("MATLAB:c_serialise:badRHS", "Bad number of RHS arguments in c_serialise");
+
+  memPtr = 0;
```

Comment:
I chose that because you may recurse and the matlab self-deserialise requires a manual increment to the pointer. It's certainly possible to pass down and up again, but it's more complicated in the logic and the pointer is guaranteed in this way to only progress.

Reply:
How do you think this would best be done?